### PR TITLE
Fix #1779: CarrierInterface additional requirement

### LINKED ISSUE #1779: CarrierInterface additional requirement
When creating a shipping module, your class should implement \Magento\Shipping\Model\Carrier\CarrierInterface

However, in order for your shipping module to actually be used, you have to implement a method called  proccessAdditionalValidation that is not required by the CarrierInterface

Failure to implement this method causes your module to be ignored in /app/code/Magento/Shipping/Model/Shipping.php because on line 250, $result will equal false because that method is not called because it does not exist in your module.

### PR DESCRIPTION
**Pull request by dverkade**
Fixed for issue #1779 as mentioned in closed PR #11293 :
- Moved methods from AbstractCarrierInterface to CarrierInterface
- Deprecated AbstractCarrierInterface
- Removed references to AbstractCarrierInterface in fthe CarrierFactoryInterface, changed them to CarrierInterface.
- Removed AbstractCarrier model implementing the AbstractCarrierInterface.
- Changed references from AbstractCarrierInterface to CarrierInterface in de Config model.

### CODE REVIEW

File: app/code/Magento/Shipping/Model/Carrier/AbstractCarrier.php [hunk 1]
```diff
@@ -17,7 +17,7 @@
  * @api
  * @since 100.0.2
  */
-abstract class AbstractCarrier extends \Magento\Framework\DataObject implements AbstractCarrierInterface
+abstract class AbstractCarrier extends \Magento\Framework\DataObject
```
Comment (omiroshnichenko) [inline app/code/Magento/Shipping/Model/Carrier/AbstractCarrier.php]: Please, revert this change because it is backward incompatible.

File: app/code/Magento/Shipping/Model/Carrier/AbstractCarrierInterface.php [hunk 2]
```diff
@@ -5,199 +5,12 @@
...
 
-use Magento\Quote\Model\Quote\Address\RateRequest;
-
 /**
  * Interface AbstractCarrierInterface
+ *
+ * @deprecated
+ * @see CarrierInterface
  */
 interface AbstractCarrierInterface
```
Comment (omiroshnichenko) [inline app/code/Magento/Shipping/Model/Carrier/AbstractCarrierInterface.php]: You must extend CarrierInterface here.

File: app/code/Magento/Shipping/Model/CarrierFactoryInterface.php [hunk 3]
```diff
@@ -5,7 +5,7 @@
  */
```
Comment (omiroshnichenko) [inline app/code/Magento/Shipping/Model/CarrierFactoryInterface.php]: Please, revert all these changes because they cause backward incompatible issue.

File: app/code/Magento/Shipping/Model/Config.php [hunk 4]
```diff
@@ -8,7 +8,7 @@
```
Comment (omiroshnichenko) [inline app/code/Magento/Shipping/Model/Config.php]: Please, revert all these changes because they cause backward incompatible issue.

**Review by omiroshnichenko — CHANGES_REQUESTED** [review]
Sync your branch with 2.3-develop.

**Comment by omiroshnichenko** [comment]
@dverkade Will you continue to work on this PR?

File: app/code/Magento/Shipping/Model/Carrier/AbstractCarrierInterface.php [hunk 5]
```diff
@@ -5,199 +5,12 @@
  */
```
Comment (omiroshnichenko) [inline app/code/Magento/Shipping/Model/Carrier/AbstractCarrierInterface.php:5]: Leave methods isStateProvinceRequired, isCityRequired, debugData in this interface. These methods must be also deprecated.

Comment (dverkade) [reply app/code/Magento/Shipping/Model/Carrier/AbstractCarrierInterface.php:5]: As described in this comment by @okorshenko the AbstractCarrierInterface should be empty
https://github.com/magento/magento2/pull/11293#issuecomment-339484781

Comment (omiroshnichenko) [reply app/code/Magento/Shipping/Model/Carrier/AbstractCarrierInterface.php:5]: As described in comment that you mentioned "We need to move all the “correct” methods to CarrierInterface", so I please you to leave unused methods in deprecated interface. When deprecated interface will be deleted, unused methods also will be removed.

**Comment by omiroshnichenko** [comment]
@dverkade Will you continue to work on this PR?

**Review by omiroshnichenko — APPROVED** [review]

**Comment by dverkade** [comment]
Thanks for chaning this, I was not able to make time last week. I see this is now approved and will be processed further.

**Comment by omiroshnichenko** [comment]
Hi @dverkade, after internal discussion, we came to that fact that we need to introduce new interface that will have correct methods, for this we must make some HLD.So I will close this PR as current solution will be backward incompatible and will not bring any benefit. Thank you for your contribution.
